contrib/gitlab/src/mill/contrib/gitlab/GitlabPublishModule.scala

```diff
@@ -101,8 +101,18 @@ trait GitlabMavenRepository {
   def repository: GitlabPackageRepository // For package discovery
 
   def mavenRepo(implicit context: Ctx): MavenRepository = {
-    val gitlabAuth = tokenLookup.resolveGitlabToken(context.log, context.env, sys.props.toMap).get
-    val auth = Authentication(gitlabAuth.headers)
-    MavenRepository(repository.url(), Some(auth))
+    val maybeRepository = tokenLookup
+      .resolveGitlabToken(context.log, context.env, sys.props.toMap)
+      .map(gitlabAuth => Authentication(gitlabAuth.headers))
+      .map(auth => MavenRepository(repository.url(), Some(auth)))
+
+    maybeRepository.getOrElse {
+      val indent = "      "
+      val tokenSearchPlaces = tokenLookup.tokenSearchOrder.mkString("", s",\n$indent", "\n")
+      val error =
+        s"Could not find Gitlab authentication for $repository, \n  Searched from:\n$indent$tokenSearchPlaces"
+      context.log.error(error)
+      throw new RuntimeException()
```

Comment:
Why not also add the error message to the exception?

Reply:
By default its a bit long and multiple lines and gets printed twice (first the log, then the exception). So when using from command line i though it was a bit spammy. 

I actually initially placed it only as exception parameter, but changed it to use context.log. Either way is fine for me, just thought that log would be bit easier to read from console.

Reply:
TBH, I really hate exceptions without any message/context. Exception with stacktraces are already ugly, and beauty should be the leading criteria for leaving them out.  There may also exist situations, where the exception pops up, but the log message isn't visible.

If it's rather the one or the other, I vote for the exception message. The fact, that Mill is currently presenting the full stack trace instead of some nicer message may change.

Reply:
Yeah me to. Would prefer to return some type that is capable of presenting error state, but this is rather close to user and that would cause at least extra .get in users build.sc.

Reply:
Instead, you could make it a task. Tasks have a proper result type. Also, you automatically get access to the correct context. 

```scala
def mavenRepo: Task[MavenRepository] = T.task {
  ...
}
```

Either just throw the exception or return a `mill.api.Result.Failure`.

https://com-lihaoyi.github.io/mill/mill/Tasks.html#_anonymous_tasks

Reply:
Just realized there are some logical duplicates. I'll do a bit larger cleanup. Hopefully on Monday (a bit busy weekend ahead).

Reply:
@aheiska Gently ping.

Reply:
Sorry for being quiet. Suddenly got a lot more work to do, so open source halted for a while. I have a some implementation that is converted to Tasks somewhat ready. From what I can recall I got a compilation error and then the other issues started pressing. Branch is not on this computer, I'll push tomorrow. It wont compile (probably something minor), but I'd like to have feedback on solution.